File: Language.Implementation.Patterns/cxx/ch02_basic_parsing_patterns/p02_lexer/lexer/Lexer.cpp
#include "Lexer.hpp"

Lexer::~Lexer() {}

void Lexer::consume() {
    p++;
    if (p >= input.length()) c = EOF;
    else c = input.at(p);
}

void Lexer::match(char x) {
    if (c == x) consume();
    else throw LexerError(std::string("expecting ") + x + "; found " + c);
}

File: Language.Implementation.Patterns/python/ch02_basic_parsing_patterns/p02_lexer_2/Lexer.py
from abc import ABC, abstractmethod

class LexerError(Exception):
    pass

class Lexer(ABC):
    EOF = -1

    def __init__(self, input):
        self.input = input
        self.p = 0
        self.c = self.input[self.p] # prime lookahead

    def consume(self):
        self.p += 1
        if self.p >= len(self.input):
            self.c = Lexer.EOF
        else:
            self.c = self.input[self.p]

    def match(self, x):
        if self.c == x:
            self.consume()
        else:
            raise LexerError("expecting " + x + "; found " + self.c)

    @abstractmethod
    def nextToken(self):
        pass

